Mathlib/Probability/Moments/SubGaussian.lean

```diff
@@ -386,6 +515,125 @@ lemma measure_ge_le (h : HasSubgaussianMGF X c μ) {ε : ℝ} (hε : 0 ≤ ε) :
+lemma measure_sum_ge_le_of_iIndepFun {ι : Type*} {X : ι → Ω → ℝ} (h_indep : iIndepFun X μ)
+    {c : ι → ℝ≥0} (h_meas : ∀ i, Measurable (X i))
+    {s : Finset ι} (h_subG : ∀ i ∈ s, HasSubgaussianMGF (X i) (c i) μ) {ε : ℝ} (hε : 0 ≤ ε) :
+    (μ {ω | ε ≤ ∑ i ∈ s, X i ω}).toReal ≤ exp (- ε ^ 2 / (2 * ∑ i ∈ s, c i)) :=
+  (sum_of_iIndepFun h_indep h_meas h_subG).measure_ge_le hε
+
+/-- **Hoeffding inequality** for sub-Gaussian random variables. -/
+lemma measure_sum_range_ge_le_of_iIndepFun {X : ℕ → Ω → ℝ} (h_indep : iIndepFun X μ) {c : ℝ≥0}
+    (h_meas : ∀ i, Measurable (X i))
+    {n : ℕ} (h_subG : ∀ i < n, HasSubgaussianMGF (X i) c μ) {ε : ℝ} (hε : 0 ≤ ε) :
+    (μ {ω | ε ≤ ∑ i ∈ Finset.range n, X i ω}).toReal ≤ exp (- ε ^ 2 / (2 * n * c)) := by
+  have h := (sum_of_iIndepFun h_indep h_meas (c := fun _ ↦ c)
+    (s := Finset.range n) (by simpa)).measure_ge_le hε
+  simpa [← mul_assoc] using h
+
+end Add
+
 end HasSubgaussianMGF
 
+section Martingale
+
+variable [StandardBorelSpace Ω]
+
+/-- If `X` is sub-Gaussian with parameter `cX` with respect to the restriction of `μ` to
+a sub-sigma-algebra `m` and `Y` is conditionally sub-Gaussian with parameter `cY` with respect to
+`m` and `μ` then `X + Y` is sub-Gaussian with parameter `cX + cY` with respect to `μ`.
+
+`HasSubgaussianMGF X cX (μ.trim hm)` can be obtained from `HasSubgaussianMGF X cX μ` if `X` is
+`m`-measurable. See `HasSubgaussianMGF.trim`. -/
+lemma HasSubgaussianMGF_add_of_HasCondSubgaussianMGF [IsFiniteMeasure μ]
+    {Y : Ω → ℝ} {cX cY : ℝ≥0} (hm : m ≤ mΩ)
+    (hX : HasSubgaussianMGF X cX (μ.trim hm)) (hY : HasCondSubgaussianMGF m hm Y cY μ) :
+    HasSubgaussianMGF (X + Y) (cX + cY) μ := by
+  suffices HasSubgaussianMGF (fun p ↦ X p.1 + Y p.2) (cX + cY)
+      (@Measure.map Ω (Ω × Ω) mΩ (m.prod mΩ) (fun ω ↦ (id ω, id ω)) μ) by
+    have h_eq : X + Y = (fun p ↦ X p.1 + Y p.2) ∘ (fun ω ↦ (id ω, id ω)) := rfl
+    rw [h_eq]
+    refine HasSubgaussianMGF.of_map ?_ this
+    exact @Measurable.aemeasurable _ _ _ (m.prod mΩ) _ _
+      ((measurable_id'' hm).prodMk measurable_id)
+  rw [HasSubgaussianMGF_iff_kernel] at hX ⊢
+  have hY' : Kernel.HasSubgaussianMGF Y cY (condExpKernel μ m)
+      (Kernel.const Unit (μ.trim hm) ∘ₘ Measure.dirac ()) := by simpa
+  convert hX.add_comp hY'
+  ext
+  rw [Kernel.const_apply, ← Measure.compProd, compProd_trim_condExpKernel]
+
+variable {Y : ℕ → Ω → ℝ} {cY : ℕ → ℝ≥0} {ℱ : Filtration ℕ mΩ}
+
+/-- Let `Y` be a random process adapted to a filtration `ℱ`, such that for all `i : ℕ`, `Y i` is
+conditionally sub-Gaussian with parameter `cY i` with respect to `ℱ (i - 1)`.
+In particular, `n ↦ ∑ i ∈ range n, Y i` is a martingale.
+Then the sum `∑ i ∈ range n, Y i` is sub-Gaussian with parameter `∑ i ∈ range n, cY i`. -/
+lemma HasSubgaussianMGF_sum_of_HasCondSubgaussianMGF [IsZeroOrProbabilityMeasure μ]
+    (h_adapted : Adapted ℱ Y) (h0 : HasSubgaussianMGF (Y 0) (cY 0) μ)
+    (h_subG : ∀ i, HasCondSubgaussianMGF (ℱ i) (ℱ.le i) (Y (i + 1)) (cY (i + 1)) μ)
```

Comment:
Could you require `h_subG` just for `i < n-1`?

Reply:
I pushed a change using `i ≤ n - 2` but I just realized that for `n = 1` what you wrote is better. Changing again.